dwi_ml/data/hdf5_creation/utils.py

```diff
@@ -0,0 +1,234 @@
+
+
+def process_streamlines(bundles_dir: pathlib.Path, bundles,
+                        header: nib.Nifti1Header, step_size: float,
+                        space: Space):
+    """Load and process a group of bundles and merge all streamlines
+    together.
+
+    Parameters
+    ----------
+    bundles_dir : pathlib.Path
+        Path to bundles folder.
+    bundles: List[str]
+        List of the bundles filenames to load.
+    header : nib.Nifti1Header
+        Reference used to load and send the streamlines in voxel space.
+    step_size: float
+        Step size to resample streamlines. If none, compress streamlines.
+    space: Space
+        Space to place the tractograms.
+
+    Returns
+    -------
+    output_tractogram : StatefulTractogram
+        All streamlines in voxel space.
+    output_lengths : List[float]
+        The euclidean length of each streamline
+    """
+
+    # Initialize
+    output_tractogram = None
+    output_lengths = []
+
+    # Find official bundles list
+    if bundles is None:
+        # Take everything found in subject bundle folder
+        bundles = [str(p) for p in bundles_dir.glob('*')]
+        if len(bundles) == 0:
+            raise ValueError("No bundle found in the boundles folder!")
+
+    for bundle_name in bundles:
+        # Load bundle
+        bundle_path = glob.glob(bundles_dir.joinpath(bundle_name + '*'))
+        if len(bundle_path) == 0:
+            raise ValueError('Bundle {} not found!'.format(bundle_name))
+        elif len(bundle_path) > 1:
+            raise ValueError(
+                'More than one file with name {}. Clean your bundles '
+                'folder.'.format(bundle_name + '*'))
+        bundle = load_tractogram(bundle_path, header)
+
+        # Send to wanted space
+        bundle.to_space(space)
+
+        # Resample or compress streamlines
+        if step_size:
+            bundle = resample_streamlines_step_size(bundle, step_size)
+            logging.debug("Resampled streamlines' step size to {}mm"
+                          .format(step_size))
+        else:
+            bundle = compress_sft(bundle)
+
+        # Compute euclidean lengths
+        output_lengths.extend(length(bundle.streamlines))
+
+        # Add processed bundle to output tractogram
+        if output_tractogram is None:
+            output_tractogram = bundle
+        else:
+            output_tractogram.streamlines.extend(bundle.streamlines)
+
+    # Transfer the streamlines to the reference space before bringing them
+    # to VOX space. NOTE: This is done in case the streamlines were tracked
+    # in a different space than the provided dataset reference
+    if output_tractogram is None:
+        output_streamlines_rasmm = []
+    else:
+        output_streamlines_rasmm = output_tractogram.streamlines
+    output_tractogram = StatefulTractogram(output_streamlines_rasmm,
+                                           dwi_ref,
+                                           space=Space.RASMM)
```

Comment:
I am not sure the best solution here. I added a dwi_ref to create the STF (error so far, not loaded anywhere). Any other solution to avoid dealing with a ref volume?

Reply:
If we're sure the output is a trk, we could put 'same' as the reference.